File: Selenium/Project/Activity8.py
"""Goal: Open the accounts page and print the contents of the table.
a. Open the browser to the Alchemy CRM site and login.
"""
#import the webdriver
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.wait import WebDriverWait

#Initialise the firefox driver
with webdriver.Firefox() as driver:
    #Declare the wait variable
    wait = WebDriverWait(driver, timeout=10)
    #Open the browser
    driver.get("http://alchemy.hguy.co/crm")
    #Find username and password elements. Pass the credential and click it
    username = driver.find_element(By.NAME, "user_name")
    password = driver.find_element(By.NAME, "username_password")
    username.send_keys("admin")
    password.send_keys("pa$$w0rd")
    driver.find_element(By.NAME, "Login").click()

    #Navigate to sales -> accounts
    driver.find_element(By.XPATH, "//a[contains(@id, 'grouptab_0')]").click()
    driver.find_element(By.XPATH, "//a[contains(@id, 'moduleTab_9_Accounts')]").click()
    
    #Find the table in the accounts web page and print the first 5 odd-numbered rows of the table
    table = wait.until(EC.visibility_of_element_located((By.XPATH, "/html/body/div[4]/div/div[3]/form[2]/div[3]/table")))
    rows = driver.find_elements(By.XPATH, "/html/body/div[4]/div/div[3]/form[2]/div[3]/table/tbody/tr")
    print("No. of rows: ", len(rows))
    count = 0
    i =0
    for r in rows:
        if(count<5):
            if(i%2==0):
                columns= r.find_elements(By.TAG_NAME,"td")
                print(columns[2].text)
                count+=1
            i+=1
    # cellValue = driver.find_element(By.XPATH, "//table[contains(@class, 'list view table-responsive')]/thead/tbody/tr[1]/td[2]")
    # print("First 5 odd-numbered rows of the table are: ", cellValue.text)

    

File: Selenium/Project/Activity8.java
/*
 1. Open the browser to the Alchemy CRM site and login
 2. Navigate to sales -> accounts page
 3. Find the table on the page and print the names of the first 5 odd numbered rows of the table to the console
 4. Close the browser
 */

package crmactivities;

import java.util.List;
import org.openqa.selenium.By;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.Test;
import org.testng.annotations.AfterClass;
import java.time.Duration;
import org.openqa.selenium.support.ui.ExpectedConditions;

public class Activity8 {

    WebDriver driver;
    WebDriverWait wait;

    @BeforeClass
    public void setUp() {
        driver = new FirefoxDriver();
        driver.get("http://alchemy.hguy.co/crm");

        WebElement userName = driver.findElement(By.name("user_name"));
        WebElement password = driver.findElement(By.name("username_password"));

        // Enter login credentials
        userName.sendKeys("admin");
        password.sendKeys("pa$$w0rd");
        driver.findElement(By.name("Login")).click();
    }

    @Test
    public void accounts() {
        // Navigate to accounts
        driver.findElement(By.xpath("//a[contains(@id, 'grouptab_0')]")).click();
        driver.findElement(By.xpath("//a[contains(@id, 'moduleTab_9_Accounts')]")).click();

        // Wait for the page to load fully and ensure table is present
        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(20));
        try {
            // Wait for any table to be present in the document
        	Thread.sleep(5000);
            WebElement table = wait.until(ExpectedConditions.presenceOfElementLocated(By.xpath("/html/body/div[4]/div/div[3]/form[2]/div[3]/table")));

            if (table != null && table.isDisplayed()) {
                System.out.println("Table found in the main document.");

                // Find rows inside tbody (Relative XPath)
                List<WebElement> rows = table.findElements(By.xpath("/html/body/div[4]/div/div[3]/form[2]/div[3]/table/tbody/tr"));
                System.out.println("rows: "+rows.size());
               
                // Loop through odd-numbered rows and print the text in each cell
              /*  int rowCount = 0;
                for (WebElement row : rows) {
                    if (rowCount % 2 != 0) { // Odd numbered rows (0-based index, odd numbers will be 1, 3, 5, etc.)
                        List<WebElement> columns = row.findElements(By.tagName("td"));
                        for (WebElement column : columns) {
                            System.out.print(column.getText() + " ");
                            //System.out.println(columns.get(0).getText());
                        }
                        System.out.println();
                    }
                    rowCount++;
                    if (rowCount >= 10) break; // Limit to first 5 odd-numbered rows
                }*/
                
                int count = 0;
                for(int i=0; i<=rows.size(); i++) {
                	if(count<5) {
                	if(i%2==0) {
                		String a= rows.get(i).findElements(By.tagName("td")).get(2).getText();
                		System.out.println(a);
                		count+=1;
                       }
                	}
                	
                }
            } else {
                System.out.println("Table not found in the main document.");
            }
        } catch (Exception e) {
            System.out.println("Error occurred while waiting for the table: " + e.getMessage());
        }
    }
    
    @AfterClass
    public void tearDown() {
        driver.quit();
    }
}
